mathesar_ui/src/components/select/Select.svelte

```diff
@@ -24,25 +24,71 @@
   export let ariaLabel: string = null;
 
   let isOpen = false;
+  let currentIndex = 0;
+  let erasedIndex = 0;
+  let t = 0;
 
   function setValue(opt: SelectOption) {
     value = opt;
     dispatch('change', {
       value,
     });
     isOpen = false;
+    erasedIndex = 0;
+    currentIndex = 0;
   }
-
+  
   function setOptions(opts: SelectOption[]) {
     if (!value && opts.length > 0) {
       setValue(opts[0]);
     }
   }
+  
+  function hoveredItem(index){
+      let items= [...document.getElementById('select-value-'+selectId).getElementsByTagName('li')];
+      if(currentIndex == items.length - 1 && index > 0){
+        currentIndex = 0;
+        index = 0;
+      }else if(currentIndex == 0 && index < 0){
+        currentIndex = items.length;
+      }else if(currentIndex == 0 && index > 0){
+        if(t <1){
+          currentIndex = -1;
+        }else if (t > 1){
+          currentIndex = 0;
+        }
+        t += 1; 
+      }
+      if(erasedIndex >= 1){
+        let hoveredElement= document.querySelector(".hovered");
+        hoveredElement.classList.remove("hovered");
+      }
+      currentIndex = currentIndex + index; 
+      items[currentIndex].classList.add("hovered");
+      erasedIndex = 1; 
+    }
```

Comment:
The algorithm to hover elements could be improved.

1. Since we are storing currentIndex, it is not essential to manipulate dom directly i.e manipulating `hovered` class.
    - Instead we can update the template to add class hovered, as follows:
      ```
       <li role='option' ... class:hovered={option[idKey] === options[currentIndex]?.[idKey]} ...>
      ```
    - We can remove all direct manipulation of dom logic that involves classList.add and classList.remove.
    - erasedIndex would not be required, when this change is done.
2. On dropdown open, the currentIndex should be set to the index of the selected item. We can dispatch an event during open/close from the dropdown for this.
3. The need for variable `t` can be avoided.

Reply:
I'm researching more about the dispatch event. I have access to the currentIndex and selected item in the Select component.  Couldn't I pass these variables as props down to the dropdown component?

Reply:
It is not a good idea to pass component specific props to different components, which are independent.

The Dropdown component is designed to be used independently, it is not specific to Select, so passing these props down would an anti-pattern.

Regarding dispatching events, you can refer the Svelte docs. https://svelte.dev/docs#createEventDispatcher

You could also refer the [Modal component](https://github.com/centerofci/mathesar/blob/master/mathesar_ui/src/components/modal/Modal.svelte), in our repo. It dispatches events upon open/close of modal.